src/runtime/contrib/tensorrt/tensorrt_module.cc

```diff
@@ -0,0 +1,182 @@
+
+  static Module LoadFromBinary(void* strm) {
+    dmlc::Stream* stream = static_cast<dmlc::Stream*>(strm);
+    std::string serialized_subgraph;
+    stream->Read(&serialized_subgraph);
+    return TensorRTModuleCreate(serialized_subgraph);
+  }
+
+ private:
+  std::string serialized_subgraph_;
+  std::unordered_map<std::string, TrtEngineAndContext> trt_engine_cache_;
+
+  // Convert TVMArgs to make compatible with VM or graph runtime.
+  std::vector<DLTensor*> ConvertInputs(tvm::TVMArgs args) {
+    std::vector<DLTensor*> inputs(args.size(), nullptr);
+    for (size_t i = 0; i < args.size(); i++) {
+      if (args[i].type_code() == kNDArrayContainer) {
+        // Relay Debug/VM uses NDArray
+        runtime::NDArray array = args[i];
+        inputs[i] = const_cast<DLTensor*>(array.operator->());
+      } else if (args[i].type_code() == kArrayHandle) {
+        // Graph runtime uses DLTensors
+        inputs[i] = args[i];
+      } else {
+        LOG(FATAL) << "Invalid TVMArgs type.";
+      }
+    }
+    return inputs;
+  }
+
+  void ExecuteEngine(const TrtEngineAndContext& engine_and_context,
+                     tvm::TVMArgs args, tvm::TVMRetValue* rv) {
+    auto engine = engine_and_context.engine;
+    auto context = engine_and_context.context;
+    const int num_bindings = engine->getNbBindings();
+    std::vector<void*> bindings(num_bindings, nullptr);
+    // Set inputs.
+    auto inputs = ConvertInputs(args);
+    const size_t num_outputs = engine_and_context.network_outputs.size();
+    CHECK_GT(inputs.size(), num_outputs);
+    // TODO(trevmorr): Assumes output is at the end - is this true?
+    for (size_t i = 0; i < inputs.size() - num_outputs; ++i) {
+      auto it = engine_and_context.network_input_map.find(i);
+      if (it != engine_and_context.network_input_map.end()) {
+        DLTensor* arg = inputs[i];
+        int binding_index = engine->getBindingIndex(it->second.c_str());
+        CHECK_NE(binding_index, -1);
+        if (!runtime::TypeMatch(arg->dtype, kDLFloat, 32)) {
+          LOG(FATAL) << "Only float32 inputs are supported.";
+        }
+        bindings[binding_index] = reinterpret_cast<float*>(arg->data);
+      }
+    }
+    // Set outputs.
+    // TODO(trevmorr): Allow multiple outputs.
+    for (size_t i = 0; i < num_outputs; ++i) {
+      const int index_in_inputs = inputs.size() - num_outputs + i;
+      DLTensor* out_arg = inputs[index_in_inputs];
+      int binding_index = engine->getBindingIndex(
+          engine_and_context.network_outputs[i].c_str());
+      CHECK_NE(binding_index, -1);
+      bindings[binding_index] = reinterpret_cast<float*>(out_arg->data);
+    }
+    // Use batch size from first input.
+    const int batch_size = inputs[0]->shape[0];
+    CHECK(context->execute(batch_size, bindings.data()))
+        << "Running TensorRT failed.";
+
+    // TODO(trevmorr): Look up bindings by name.
+    // TODO(trevmorr): Allow multiple outputs.
+    *rv = bindings[num_bindings - num_outputs];
```

Comment:
We may want to assert and raise error messages when there are multiple outputs. We could encode outputs in to args as well. We also support runtime object with multiple entries using ADT

Reply:
Thanks Zhi for reviewing!
I added a test case for multiple outputs. I am currently using the args for outputs and it seems to be working with valid results.
I am still confused about what `*rv` should be pointing to - but it appears it doesn't matter based on these results.